Added code review pattern (#536)
UKHomeOffice/engineering-guidance-and-standards: docs/patterns/code-reviews.md

### Point out the good things as well as bad

Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.

### Have a pull request template

A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code. You can find an [example in our engineering guidance and standards repository](https://github.com/UKHomeOffice/engineering-guidance-and-standards/blob/main/.github/PULL_REQUEST_TEMPLATE.md).

### Support your code reviews with automation

Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.

---

## Considerations

### Some of the things that should be checked for

> This is not an exhaustive list, just some recommendations, but your team should have their own standards. A pull request template can help with this by creating a checklist of what to look for.

Reviewer: This is not an exhaustive list, just some recommendations.  Your team should ...

Does it work as intended?
: Ensuring that code works as intended is a crucial part of the code review process because it verifies functionality, preventing bugs and errors from reaching production, which saves time and resources. It also ensures a smooth and reliable user experience, maintaining user trust and satisfaction.

Is it tested?

> : Testing code is important, make sure to check that current tests are passed and make suggestions where further automations would help

Reviewer: Just needs a full stop on the end here.

Is it easy to read?
: Readability is important because it makes the code easier to understand, refactor, maintain and debug ensuring that developers can work on it effectively in the future.

Is the software well designed?
: Checking for things such as good error handling and logging can help developers debugging and provide a better user experience. Making sure the code fits with the rest of the project is also important. Finally checking if they have missed any edge cases or you can see any performance issues or bugs.

Does the code protect against misuse?
: Protecting against security vulnerabilities is crucial, checking for things such as allowing code injection, not validating user inputs or including Personally Identifiable Information (PII) in error messages will help to catch problems early.

Do the changes consider existing code?

> : Reusability is important because it enhanced efficiency and minimises errors and leads to faster development and makes code easier to maintain. Frameworks, tools and standards within a team can also help a developer reuse functions they provide, make sure to check that code can't make use of these sources. Pay particular attention to this if the code may have been generated by AI which is not aware of shared functions/services etc.

Reviewer: There is a typo here: 
Reusability is important because it **enhances** efficiency

Is the code documented

> : Documentation ensures that teams can more easily maintain and developer a codebase further. Checking that you're happy with the level of documentation, especially if you were to work on this code in the future.

Reviewer: maintain and develop a ...

### Point out the good things as well as bad

Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.

### Have a pull request template

A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code.

### Support your code reviews with automation

Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.

---

## Considerations

### Some of the things that should be checked for

This is not an exhaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.

Does it work as intended?
: Does the code look like it would work like the author intended? Does it work as users would expect?

Is it tested?

> : Are the tests well designed? Is there enough coverage? Are the tests automated? Does it pass automated testing? Does it pass linting checks?

Reviewer: I think we could reformat this instead of a set of questions.  As a pattern we should suggest some good practices for code reviews.
Something like "Testing in code is important, make sure to check that current tests are passed and make suggestions where further automations would help"

---
layout: pattern
order: 1
title: Code reviews
date: 2025-04-02
tags:
  - Software design
  - Ways of working
---

Code reviews are a vital part of the software lifecycle. They can keep code quality to a high standard as well as share knowledge between team members.

---

## Solution

### Code reviews should be part of your teams workflow

Time should be allocated as part of your sprint to review code, as it's important to keep high coding standards.

### Code reviews should be constructive and friendly

When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.

### Point out the good things as well as bad

Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.

### Have a pull request template

> A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.

Reviewer 2: ```suggestion
A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code.
```
Probably don't need both "also" + "too" in this last sentence.

### Support your code reviews with automation

Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.

---

## Considerations

### Some of the things that should be checked for

> This is not an exaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.

Reviewer 3: ```suggestion
This is not an exhaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
```

Does it work as intended?
: Does the code look like it would work like the author intended? Does it work as users would expect?

Is it tested?
: Are the tests well designed? Is there enough coverage? Are the tests automated? Does it pass automated testing? Does it pass linting checks?

Is it easy to read?

> : Can you clearly understand what the code is doing? Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?

Reviewer 2: Small wording suggestion

```suggestion
: Can you clearly understand what the code is doing? Do variables, functions and classes have descriptive names? Is the code too complex? Are comments useful and clear? Are the comments needed?
```

Is the software well designed?

> : Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any security vulnerabilities, performance issues, or other bugs?

Reviewer 2: Security vulnerabilities are covered in the point below, so maybe don't need it in this one as well.

Also, may make sense to consider logging alongside error handling?

```suggestion
: Is there good error handling and logging? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any performance issues or other bugs?
```

Does the code protect against misuse?
: Does the code have any security vulnerabilities, e.g. does it allow code injection? Do error messages contain too much information? Is user input validated?

Do the changes consider existing code?

> : Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and convetions of the team. Does it make good use of the application's framework?

Reviewer 3: ```suggestion
: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team. Does it make good use of the application's framework?
```

> : Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team. Does it make good use of the application's framework?

Reviewer 2: ```suggestion
: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team? Does it make good use of the application's framework?
```

>   Pay particular attention to this if the code may have generated by AI which is not aware of shared functions/services etc?

Reviewer 2: ```suggestion
  Pay particular attention to this if the code may have been generated by AI which is not aware of shared functions/services etc.
```

Is the code documented?
: Should there be more documentation around this code? Would you need more information before working on it?

Does it follow standards?
: Does it follow the team's standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?

> Do regular expessions make sense?

Reviewer 3: ```suggestion
Do regular expressions make sense?
```

: Does a regex work as expected? Does it include all edge cases?

> 
> ---

Reviewer 2: ```suggestion
```
Not sure if it's needed but this section break makes it look like there's an empty row at the bottom of the table, similar to the bottom of https://engineering.homeoffice.gov.uk/patterns/docs-as-code/ 

Seems to render fine when removed locally - but again, not sure if it's needed for something else.

### Code reviews should be constructive and friendly

When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.

### Point out the good things as well as bad

Positive feedback can help build a positive culture within your team and it's always nice to hear positive feedback.

### Have a pull request template

A pull request template can help you create a check list of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.

### Support your code reviews with automation

Automate as much as possible to help relive the code review burdon, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.

---

## Considerations

### Some of the things that should be checked for

This is not an exaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.

> #### Does it work as attended?

Reviewer 4: ```suggestion
#### Does it work as intended?
```

---
layout: pattern
order: 1
title: Code reviews
date: 2025-03-21
tags:
  - Software design
  - Ways of working
---

Code reviews are a vital part of the software lifecycle. They can keep code quality to a high standard as well as share knowledge between team members.

---

## Solution

### Code reviews should be part of your teams workflow

Time should be allocated as part of your sprint to review code, as it's important to keep high coding standards.

### Code reviews should be constructive and friendly

When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.

### Point out the good things as well as bad

> Positive feedback can help build a positive culture within your team and it's always nice to hear positive feedback.

Reviewer 3: ```suggestion
Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.
```

### Have a pull request template

> A pull request template can help you create a check list of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.

Reviewer 3: ```suggestion
A pull request template can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
```

> A pull request template can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.

Reviewer 3: ```suggestion
A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too. 
```

### Support your code reviews with automation

> Automate as much as possible to help relive the code review burdon, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.

Reviewer 3: ```suggestion
Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.
```

---

## Considerations

### Some of the things that should be checked for

This is not an exaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.

#### Does it work as intended?

Does the code look like it would work like the author intended? Does it work as users would expect?

#### Is it tested?

Are the tests well designed? Is there enough coverage? Are the tests automated?

#### Is it easy to read?

> Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?

Reviewer 3: ```suggestion
#### Is it easy to read?

Can you clearly understand what the code is doing? Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
```

#### Is the software well designed and bug free?

Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?

> #### Could the code be used for something other than the intended purpose?
> 
> Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?

Reviewer 3: ```suggestion
#### Does the code protect against misuse?

Does the code have any security vulnerabilities, e.g. does it allow code injection? Do error messages contain too much information? Is user input validated?
```

> #### Does any of the code duplicate an existing function?
> 
> Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?

Reviewer 3: ```suggestion
#### Do the changes fit well into the existing code?
Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and convetions of the team. Does it make good use of the application's framework?

Pay particular attention to this if the code may have generated by AI which is not aware of shared functions/services etc?
```

#### Is the code documented?

Should there be more documentation around this code? Would you need more information before working on it?

#### Does it follow standards?

> Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?

Reviewer 3: ```suggestion
Does it follow the team's standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
```

> #### Does it work as intended?
> 
> Does the code look like it would work like the author intended? Does it work as users would expect?
> 
> #### Is it tested?
> 
> Are the tests well designed? Is there enough coverage? Are the tests automated?
> 
> #### Is it easy to read?
> 
> Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
> 
> #### Is the software well designed and bug free?
> 
> Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?
> 
> #### Could the code be used for something other than the intended purpose?
> 
> Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?
> 
> #### Does any of the code duplicate an existing function?
> 
> Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?
> 
> #### Is the code documented?
> 
> Should there be more documentation around this code? Would you need more information before working on it?
> 
> #### Does it follow standards?
> 
> Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
> 
> #### Do the automated tests pass?
> 
> Does it pass automated testing? Does it pass linting checks?
> 
> #### Do regular expessions make sense?
> 
> Does a regex work as expected? Does it include all edge cases?

Reviewer 3: Does this work better as a definition/summary list?

```suggestion
Does it work as intended?
: Does the code look like it would work like the author intended? Does it work as users would expect?

Is it tested?
: Are the tests well designed? Is there enough coverage? Are the tests automated?

Is it easy to read?
: Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?

Is the software well designed and bug free?
: Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?

Could the code be used for something other than the intended purpose?
: Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?

Does any of the code duplicate an existing function?
: Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?

Is the code documented?
: Should there be more documentation around this code? Would you need more information before working on it?

Does it follow standards?
: Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?

Do the automated tests pass?
: Does it pass automated testing? Does it pass linting checks?

Do regular expessions make sense?
: Does a regex work as expected? Does it include all edge cases?
```

![image](https://github.com/user-attachments/assets/3579448f-6cc0-4be0-a689-0c2a2b57bdb3)

> Does it pass automated testing? Does it pass linting checks?

Reviewer 3: ```suggestion
Does it pass automated testing? Does it pass linting checks? Have tests been added to validate new functionality?
```
